tests/upgrade_path/test_upgrade_path.py

```diff
@@ -0,0 +1,198 @@
+import pytest
+import logging
+import json
+import os
+from common import reboot
+from jinja2 import Template
+import ipaddr as ipaddress
+from ptf_runner import ptf_runner
+from common.platform.ssh_utils import prepare_testbed_ssh_keys
+from datetime import datetime
+
+from common.fixtures.ptfhost_utils import copy_ptftests_directory   # lgtm[py/unused-import]
+from common.fixtures.ptfhost_utils import change_mac_addresses      # lgtm[py/unused-import]
+
+pytestmark = [
+    pytest.mark.topology('any')
+]
+
+logger = logging.getLogger(__name__)
+
+SOURCE_UPGRADE_SCRIPT = 'upgrade_path/upgrade_sonic.sh'
+UPGRADE_SCRIPT = '/tmp/upgrade_sonic.sh'
+
+TMP_VLAN_PORTCHANNEL_FILE = '/tmp/portchannel_interfaces.json'
+TMP_VLAN_FILE = '/tmp/vlan_interfaces.json'
+TMP_PORTS_FILE = '/tmp/ports.json'
+
+
+@pytest.fixture(scope="module")
+def duthost_original_image(duthost):
+    original_image = ""
+    res = duthost.shell('sonic_installer list', module_ignore_errors=True)
+    for line in res['stdout'].splitlines():
+        if line.startswith('Current:'):
+            original_image = line.split(':')[-1].strip()
+            break
+    logger.info("Original image is {}".format(original_image))
+    return original_image
+
+
+@pytest.fixture(scope="module")
+def setup(localhost, ptfhost, duthost, duthost_original_image):
+    prepare_ptf(ptfhost, duthost)
+    yield
+    cleanup(localhost, ptfhost, duthost, duthost_original_image)
+
+
+def prepare_dut(duthost):
+    # copy script to dut
+    duthost.copy(src=SOURCE_UPGRADE_SCRIPT,
+                 dest=UPGRADE_SCRIPT,
+                 mode=0755)
+
+
+def cleanup(localhost, ptfhost, duthost, duthost_original_image):
+    original_image = duthost_original_image
+    logger.info("Preparing to cleanup and restore to {}".format(original_image))
+    # restore orignial image
+    logger.info("Set default image to {}".format(original_image))
+    # try to restore original image
+    duthost.shell("sonic_installer set_default {}".format(original_image),
+                  module_ignore_errors=True)
```

Comment:
How do you guarantee that the original image is still there? Since this is upgrade path testing, assume originally there are two images on the SONiC switch:
* IMAGE_A
* IMAGE_B

After upgrade to IMAGE_C, then the images on switch:
* IMAGE_C
* IMAGE_A

After upgrade to IMAGE_D, then the images on switch:
* IMAGE_D
* IMAGE_C.

Now if you try to run `sonic_installer set_default <IMAGE_A>`, it won't work.

Reply:
This is true. We could potentially have 3 images:

1. The original image.
2. The base image.
3. The target image.

When the target image is not the same as original image, the original image could be wiped. We need to also pass in an original image url for restoring.

Reply:
Yes. A new parameter restore_to_image is added.
Thanks!